src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -2000,105 +1996,3 @@ QueryRteIdentities(Query *queryTree)
 
 	return queryRteIdentities;
 }
-
-
-/*
- * RemoveDuplicateJoinRestrictions gets a join restriction context and returns a
- * newly allocated join restriction context where the duplicate join restrictions
- * removed.
- *
- * Note that we use PostgreSQL hooks to accumulate the join restrictions and PostgreSQL
- * gives us all the join paths it tries while deciding on the join order. Thus, for
- * queries that has many joins, this function is likely to remove lots of duplicate join
- * restrictions. This becomes relevant for Citus on query pushdown check peformance.
- */
-JoinRestrictionContext *
-RemoveDuplicateJoinRestrictions(JoinRestrictionContext *joinRestrictionContext)
```

Comment:
It's easy to see that the overhead is due to us calling `list_difference` (via `LeftListIsSubset`) repeatedly. This is because, calling `list_difference` is an overhead itself.

The first idea that came to my mind was to convert each `joinRestriction->joinRestrictionInfoList` to a set in `multi_join_restriction_hook` but would also be a costly operation.

Other than that, the only function could benefit from removing duplicate join restrictions was `GenerateAttributeEquivalencesForJoinRestrictions` as it's complexity O(n^2) where n is the number of join restrictions.

However, as can be seen in pr description, `RemoveDuplicateJoinRestrictions` doesn't seem to be paying that cost.

----

Also, from the memory perspective, we were already not `free`ing duplicate join restrictions, so we would not have additional memory overhead if we remove `RemoveDuplicateJoinRestrictions`

Reply:
The loop inside RemoveDuplicateJoinRestrictions()  become  something like:
```
foreach(joinRestrictionCell, joinRestrictionContext->joinRestrictionList)   // RemoveDuplicateJoinRestrictions
          foreach_ptr(joinRestrictionInContext, joinRestrictionInContextList) // ContextCoversJoinRestriction
                foreach(cell, list1)   // list_difference
                    foreach(cell, list)   //  list_member
```

which is very dangerous as I noted in the  benchmarks above.


Overall, the planning times are extremely high for queries  with more than 50 tables, but I guess  that's a problem to be solved later